Master node disconnect

joshua-adams-1 opened the pull request:
Extends the Coordinator so that we don't prematurely close the connection to a joining node. This prevents a `node-join: [{}] with reason [{}]; for troubleshooting guidance, see {}` `WARN` log being emitted unnecessarily. 

Closes #126192

Jira Ticket - ES-11449

joshua-adams-1 committed:
Don't fail on first node-join publication failure

Extends the `Coordinator.handleJoinRequest` `onFailure` method in the
callback to only fail if the node is not in already in the
`ClusterState`. This avoids a rare corner case where a master node's
`ClusterState` already has a node in it that is attempting to join the
cluster, and it logs an incorrect error message

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Checkstyle

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Adds Node Joining Integration Tests

Adds two new integration test suites, `NodeJoiningIT` and
`NodeJoiningMasterElectionIT`. These included tests related to the
coordinator logic when a node joins the cluster.

There is also a proposed solution in the Coordinator class, currently
commented out. This will be uncommented in a follow up commit

joshua-adams-1 committed:
Adds nodeExistsWithName method

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Minor tweaks

pxsalehi commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,456 @@
+
+        try {
+            ensureSufficientMasterEligibleNodes();
+            DiscoveryNode masterNode = internalCluster().clusterService().state().nodes().getMasterNode();
```
Regarding test flakiness, one contributor could be where you get the info that you use in your assertions. Using `internalCluster().clusterService()` doesn't necessarily give you the most up to date value since it might not fetch the instance from the current master node. This is documented on that method, but want to make sure you have consider this random behaviour that this utility can have. You can fetch the master node instance by providing the master node name or similar to the test in `TransportMasterNodeActionIT`, which I assume this is partially based on, use a master client.

pxsalehi replied:
Having tried that out briefly, I think the test failure is on longer due to the latch you mentioned but rather log expectation assertions. Hope that hleps!

pxsalehi replied:
(at least based on 20-something runs that used to get 8-10 failures in the latch check, which now only fails once later at the log expectation assertion)

joshua-adams-1 committed:
Uncomments out solution, and makes NodeJoiningIT
test that the warn log is not present

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Ran ./gradlew spotlessApply precommit

DaveCTurner reviewed (commented):
Unclear if this is ready for review or not - it's still marked as a draft but you've requested reviews. 

I note that you're using force-push rather than merge:

<img width="608" height="180" alt="image" src="https://github.com/user-attachments/assets/49f656f5-1664-4d89-afd5-7c82572a584e" />

Please don't do that once other folks have started to look at a PR unless absolutely unavoidable. It makes it much harder to track the history of changes against the comments.

joshua-adams-1 commented:
> Unclear if this is ready for review or not - it's still marked as a draft but you've requested reviews.
> 
> I note that you're using force-push rather than merge:
> 
> <img alt="image" width="608" height="180" src="https://private-user-images.githubusercontent.com/5058284/476942940-49f656f5-1664-4d89-afd5-7c82572a584e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ5OTA3MTksIm5iZiI6MTc1NDk5MDQxOSwicGF0aCI6Ii81MDU4Mjg0LzQ3Njk0Mjk0MC00OWY2NTZmNS0xNjY0LTRkODktYWZkNS03YzgyNTcyYTU4NGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MTJUMDkyMDE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWI5ZTE3ZjkwN2I5NzZmNzk2NTc1ZTViNzgwMzRjZWY2ZWQwZGNjMjE2NTU4Njc0NzNmYTM3NzEwMjhhMGQ3NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.X43rvAJmAKBN-IJBr7b-uIW9TFty0kd-CKfgPxxdoGc">
> Please don't do that once other folks have started to look at a PR unless absolutely unavoidable. It makes it much harder to track the history of changes against the comments.

Hey! Sorry, no I wasn't expecting a review yet. I added you as a reviewer, but I didn't realise that action sent it for review, because there's a button at the bottom of my PR saying `Ready for Review`. I was under the impression clicking *that* sent it for review. Still learning, sorry! Thank you for the comments, I'll get it cleaned up. And I shall stop using force merges!

joshua-adams-1 committed:
Modify Coordinator Logic

- Removes MasterElectionTestCase
- reverts TransportMasterNodeActionIT
- Modifies NodeJoiningIT to use ClusterStateUtils
rather than CountDownLatches
- Modifies the Coordinator to keep the connection open
only when there is a FailedToCommitClusterStateException
and until the next committed cluster state update

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -662,11 +669,41 @@ private void handleJoinRequest(JoinRequest joinRequest, ActionListener<Void> joi
+                     */
+                    ClusterStateListener listener = new ClusterStateListener() {
+                        @Override
+                        public void clusterChanged(ClusterChangedEvent event) {
+                            // Now it's safe to close the connection
+                            Releasables.close(response);
+                            // Remove this listener to avoid memory leaks
+                            clusterService.removeListener(this);
```
rather than using `ActionListener#runBefore` you'd typically capture the exception with `ActionListener#delegateResponse`. But in this case we're already capturing the success response with `ActionListener#delegateFailure` so it's probably best to just create a completely `new ActionListener` that does the right thing with both success and exceptional responses.

joshua-adams-1 replied:
Updated in a668ab10a80052028fc015684fe78a3339c12480

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Generalise logging expectations

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 commented:
As a note, I got the error message:
```
! [rejected] name-of-your-branch -> name-of-your-branch (non-fast-forward)
error: failed to push some refs to 'https://github.com/YourGitHubHandle/kibana.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
and was required to `force-push` just this once, as recommended [here](https://www.elastic.co/docs/extend/kibana/development-github)

DaveCTurner commented:
We'll survive, but in general it's better to adjust your own local state to be compatible with the history on the remote rather than to force-push like that. A `git pull && git merge` is typically all it needs, although you can get fancy with rebasing your local commits on top of the remote head if you'd prefer.

Commit:
[CI] Auto commit changes from spotless

smalyshev committed:
Add YAML test for "missing lookup key" scenario (#132870)

joshua-adams-1 commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,563 @@
+    private String generateNodeDescriptionForNewDiscoveryNode(int numberOfNodesOriginallyInCluster, DiscoveryNode masterNode) {
+        // Nodes are named `node_s0`, `node_s1` etc ...
+        // Therefore, if there are N nodes in the cluster, named `node_s0` ... `node_sN-1`, N+1 will be named `node_sN`
+        String newNodeName = "node_s" + numberOfNodesOriginallyInCluster;
```
Good point, I have updated in cc67155c8cd61f842a6ce15c75bf3f07363490fd

joshua-adams-1 commented on server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -267,6 +267,21 @@ public boolean nodeExists(DiscoveryNode node) {
+     * @param name of the node which existence should be verified
+     * @return <code>true</code> if the node exists. Otherwise <code>false</code>
+     */
+    public boolean nodeExistsWithName(String name) {
```
Updated in a668ab10a80052028fc015684fe78a3339c12480

joshua-adams-1 commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/MasterElectionTestCase.java:
```diff
@@ -0,0 +1,124 @@
+     * @param cleanupTasks The list of cleanup tasks
+     * @return A latch that will be released when the master acknowledges it's re-election
+     */
+    protected CountDownLatch configureElectionLatchForReElectedMaster(
```
Updated in a668ab10a80052028fc015684fe78a3339c12480

JonasKunz committed:
Add memory accounting to exponential histogram library. (#132580)


---------

Co-authored-by: elasticsearchmachine <infra-root+elasticsearchmachine@elastic.co>
Co-authored-by: Kostas Krikellas <131142368+kkrik-es@users.noreply.github.com>

iverase committed:
Vectorize BQVectorUtils#packAsBinary (#132923)

Vectorize by using a helper array that defines the shifts we need to apply to the vector elements.

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.qa.mixed.EsqlClientYamlIT test {p0=esql/230_folding/CIDR_MATCH with non-foldable CIDRs} #132931

idegtiarenko committed:
Remove mutes for resolved CsvTests issues (#132924)

carlosdelest committed:
ESQL - Allow null values in vector similarity functions (#132919)

* Allow null values and return null as result

* Fix indentation

* [CI] Auto commit changes from spotless

* Remove unnecessary return

* Fix verifier tests

* Fix bwc tests introducing capability

---------

Co-authored-by: elasticsearchmachine <infra-root+elasticsearchmachine@elastic.co>

iverase committed:
Vectorize BQSpaceUtils#transposeHalfByte (#132935)

DiannaHohensee committed:
Send max of two types of max queue latency to ClusterInfo (#132675)

The TransportNodeUsageStatsForThreadPoolsAction now takes the max
latency of any task currently queued in the write thread pool queue AND
the previously collected max queue latency of any task dequeued since
the last call. This covers the possibility that queue times can rise
greatly before being reflected in execution: imagine all the write
threads are stalled or have long running tasks. This action feeds a max
queue latency stat to the ClusterInfo. Follow up from ES-12233.

Adds additional IT testing to exercise both forms of queue latency, a
followup for ES-12316.

-------------

Completing the follow up testing [Henning requested
previously](https://github.com/elastic/elasticsearch/pull/131480/files#r2243610807).

iverase committed:
[DiskBBQ] Replace n_probe, related to the number of centroids with  visit_percentage, related to the number of documents (#132722)

This commit changes the way we budget how many vectors we are going to score during search.

jonathan-buttner committed:
[ML] Add spec files for Llama and AI21 (#132724)

* Adding missing specs

* Update rest-api-spec schema to accept digits

---------

Co-authored-by: Sylvain Wallez <sylvain@elastic.co>

smalyshev committed:
Remove awaits for closed issues (#132306)

* Remove awaits for closed issues

mark-vieira committed:
Suppport per-project behavior in ESQL extra verifiers (#131884)

parkertimmins committed:
Add random tests with match_only_text multi-field (#132380)

Add tests to the logsdb randomized tests which use multi-fields for each of the text, keyword, match_only_text, and wildcard types. For each type allow each other type of multi-field.

jordan-powers committed:
Store ignored source in unique stored fields per entry (#132142)

This PR does the following:
* Stores each _ignored_source entry in a unique stored field called
  _ignored_source.<field_name>
* Coalesces multiple entries for the same field name into a single lucene
  stored field
* Adds the WildcardFieldMaskingReader so that when running synthetic source
  roundtrip tests, we can ignore differences in fields that match the
  pattern ignored_source.*

For now, these changes are by default disabled behind a feature flag.

smalyshev committed:
Rename skipping logic to remove hard link to skip_unavailable (#132861)

elasticsearchmachine committed:
Mute org.elasticsearch.index.mapper.LongFieldMapperTests testFetchMany #132948

masseyke committed:
Adding simulate ingest effective mapping (#132833)

joegallo committed:
Precompute the BitsetCacheKey hashCode (#132875)

julian-elastic committed:
Fix failing UT by adding a required capability (#132947)

Fix failing UT by adding a required capability

elasticsearchmachine committed:
Mute org.elasticsearch.index.mapper.LongFieldMapperTests testFetch #132956

JeremyDahlgren committed:
Remove CrossClusterCancellationIT.createLocalIndex() (#132952)

createLocalIndex() was added in #121240 but is unused now
after testCloseSkipUnavailable() was removed in #123054.

masseyke committed:
Unmuting simulate index data stream mapping overrides yaml rest test (#132946)

elasticsearchmachine committed:
Mute org.elasticsearch.cluster.ClusterInfoServiceIT testMaxQueueLatenciesInClusterInfo #132957

smalyshev committed:
Introduce execution location marker for better handling of remote/local compatibility (#132205)

* Introduce execution location marker for better handling of remote/local compatibility

svilen-mihaylov-elastic committed:
Implement v_magnitude function (#132765)

Implements v_magnitude ESQL function

nicktindall committed:
Breakdown undesired allocations by shard routing role (#132235)

In order that we can prevent scale-down in stateless when there are undesired allocations specifically in the indexing tier

Closes: ES-12221

Co-authored-by: Pooya Salehi <pxsalehi@users.noreply.github.com>

gbanasiak committed:
Switch to PR-based benchmark pipeline defined in ES repo (#132941)

elasticsearchmachine committed:
Mute org.elasticsearch.test.rest.yaml.CcsCommonYamlTestSuiteIT test {p0=search/400_synthetic_source/_doc_count} #132965

DiannaHohensee committed:
Implement WriteLoadConstraintDecider#canAllocate (#132041)

The initial version of the write load decider with #canAllocate
implemented. Checks whether the new node assignment for a shard
would exceed the node's simulated utilization threshold.

Closes ES-12564

idegtiarenko committed:
Simplify EsqlSession (#132848)

elasticsearchmachine committed:
Mute org.elasticsearch.index.mapper.LongFieldMapperTests testSyntheticSourceWithTranslogSnapshot #132964

dnhatn committed:
Speed up loading keyword fields with index sorts (#132950)

Reading keyword fields that are the primary sort in the index can be sped up by skip reading, as identical values are stored together. In this case, we can use values from the doc_values skipper instead of loading values from doc_values. However, the doc_values skipper is not enabled yet. Here, we use two buffers when reading ordinals: one for the beginning of the block and one for the end. If both return the same value, we can skip the middle. There is a follow-up step where we fill the values in the middle until we reach the last value. This optimization should speed up time-series queries.

joshua-adams-1 committed:
Merge remote-tracking branch 'upstream/main' into master-node-disconnect

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Remove logger.info

DaveCTurner reviewed (commented):
Thanks Josh, I did a first pass and left some comments.

joshua-adams-1 committed:
David Turner Comments

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Fix unit tests

joshua-adams-1 committed:
Merge branch 'master-node-disconnect' of github.com:joshua-adams-1/elasticsearch into master-node-disconnect

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,308 @@
+            .getDataNodes()
+            .size();
+        // Determine upfront who we want the next master to be
+        final var newMasterNodeName = nodeNames.getFirst();
```
This will sometimes re-elect the original master, causing the `assertNotEquals(originalMasterNodeName, internalCluster().getMasterName());` lower down to fail.

I'm not sure that's a bad thing - IMO it'd be simpler to have one test that covers both cases, but if you do want to keep them separate then you need to do something like this:

```suggestion
        final var newMasterNodeName = randomValueOtherThan(originalMasterNodeName, () -> randomFrom(nodeNames));
```

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,308 @@
+            .nodes()
+            .getMasterNodes()
+            .size();
+        int numberOfDataNodesOriginallyInCluster = internalCluster().clusterService(originalMasterNodeName)
```
Do we need to check data nodes specifically here? I think just counting the number of nodes is fine.

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,308 @@
+
+        long originalTerm = internalCluster().clusterService(masterNodeName).state().coordinationMetadata().term();
+        int numberOfNodesOriginallyInCluster = internalCluster().clusterService(masterNodeName).state().getNodes().size();
+        int numberOfMasterNodesOriginallyInCluster = internalCluster().clusterService(masterNodeName)
+            .state()
+            .nodes()
+            .getMasterNodes()
+            .size();
+        int numberOfDataNodesOriginallyInCluster = internalCluster().clusterService(masterNodeName).state().nodes().getDataNodes().size();
```
Likewise here, I think we just need to count the nodes, no need to distinguish them by role.

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,308 @@
+            assertEquals(masterNodeName, internalCluster().getMasterName());
+            logger.info("Master has been re-elected");
+
+            try {
+                // Await for N to be in the cluster state of all nodes
+                for (String nodeName : namesOfAllNodesInOriginalCluster) {
+                    ClusterServiceUtils.awaitClusterState(
+                        logger,
+                        clusterState -> nodeExistsWithName(clusterState.nodes(), newNodeName),
+                        internalCluster().clusterService(nodeName)
+                    );
+                }
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
```
This should already be true before `startDataOnlyNode` returns, no need to wait. We could check it's true if you'd like, but for that I'd recommend iterating over `internalCluster().getInstances(ClusterService.class)` rather than looking each one up by node name.

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,308 @@
+            logger.info("Master publishing ban removed");
+
+            // Wait until the master acknowledges its re-election. The master is only re-elected once it's publishing ban is lifted
+            SubscribableListener<Void> masterKnowsItsReElectedListener = ClusterServiceUtils.addTemporaryStateListener(
+                internalCluster().getInstance(ClusterService.class, masterNodeName),
+                clusterState -> {
+                    DiscoveryNode currentMasterNode = clusterState.nodes().getMasterNode();
+                    long currentTerm = clusterState.coordinationMetadata().term();
+                    return currentMasterNode != null && currentMasterNode.getName().equals(masterNodeName) && currentTerm > originalTerm;
+                }
+            );
+            safeAwait(masterKnowsItsReElectedListener);
```
This should also be true before `startDataOnlyNode` returns, no need to wait.

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,308 @@
+                "warn message with troubleshooting link",
+                "org.elasticsearch.cluster.coordination.NodeJoinExecutor",
+                Level.WARN,
+                "node-join: [*] with reason ["
+                    + ReferenceDocs.UNSTABLE_CLUSTER_TROUBLESHOOTING
+                    + "]; for troubleshooting guidance, see "
+                    + "https://www.elastic.co/docs/troubleshoot/elasticsearch/troubleshooting-unstable-cluster*"
```
I don't think we _ever_ emit this message, the `reason` would be `disconnected` rather than a troubleshooting docs URL. `UnseenEventExpectation` assertions are tricky, it's easy to get them to pass just by being overly specific like this. You need to be sure that these tests fail without the production code change, which would have caught this. In practice I think we can just match `*` for the message, we should see no warnings at all here.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -665,7 +671,47 @@ public void onResponse(Releasable response) {
+
+                                clusterService.addListener(clusterStateListener);
+                            } else {
+                                Releasables.close(response);
+                                joinListener.onFailure(e);
```
This is ok as long as `Releasables.close(response)` doesn't throw, but `ignoredListener.onFailure(e)` would be better since we already combined `response` and `joinListener` into that listener in the `runBefore` line above. That way we're making sure that every listener in scope is complete, and also not duplicating the `Releasables.close(response)`. Similarly on the success path within the listener.

These listeners are all a bit of a tangle now tho - I think it might simplify things a bit if we rewrote it in terms of promises, WDYT?

```java
            public void onResponse(Releasable response) {

                SubscribableListener
                    // validate the join request: can the remote node deserialize our cluster state and does it respond to pings?
                    .<Void>newForked(l -> validateJoinRequest(joinRequest, l))

                    // add the joining node to the cluster state
                    .<Void>andThen(l -> processJoinRequest(joinRequest, l.delegateResponse((ll, e) -> {
                        if (e instanceof FailedToCommitClusterStateException) {
                            // The commit failed (i.e. master is failing over) but this does not imply that the join has actually failed:
                            // the next master may have already accepted the state that we just published and will therefore include the
                            // joining node in its future states too. Thus we need to wait for the next committed state before we know the
                            // eventual outcome, and we need to wait for that before we can release (our ref to) the connection and complete
                            // the listener.

                            // NB we are on the master update thread here at the end of processing the failed cluster state update, so this
                            // all happens-before any cluster state update that re-elects a master and thus the listener added below will
                            // definitely see the next cluster state update.
                            assert ThreadPool.assertCurrentThreadPool(MasterService.MASTER_UPDATE_THREAD_NAME);

                            clusterService.addListener(new ClusterStateListener() {
                                @Override
                                public void clusterChanged(ClusterChangedEvent event) {
                                    // Keep the connection open until the next committed state
                                    final var discoveryNodes = event.state().nodes();
                                    if (discoveryNodes.getMasterNode() != null) {
                                        // Remove this listener to avoid memory leaks
                                        clusterService.removeListener(this);

                                        if (discoveryNodes.nodeExists(joinRequest.getSourceNode().getId())) {
                                            ll.onResponse(null);
                                        } else {
                                            ll.onFailure(e);
                                        }
                                    }
                                }
                            });
                        } else {
                            ll.onFailure(e);
                        }
                    })))

                    // Whatever the outcome, release (our ref to) the connection we just opened and notify the joining node.
                    .addListener(ActionListener.runBefore(joinListener, () -> Releasables.close(response)));
            }
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -665,7 +671,47 @@ public void onResponse(Releasable response) {
+                                            Releasables.close(response);
+                                            // Remove this listener to avoid memory leaks
+                                            clusterService.removeListener(this);
+                                            joinListener.onResponse(null);
```
Maybe we should check to see if the joining node is in the new cluster state, and fail the listener with `e` if not.

joshua-adams-1 committed:
Comments

joshua-adams-1 committed:
Unused method

joshua-adams-1 committed:
Merge branch 'master-node-disconnect' of github.com:joshua-adams-1/elasticsearch into master-node-disconnect

joshua-adams-1 commented:
Thank you David for your help on this! Appreciate there's been a bit of churn

Commit:
[CI] Auto commit changes from spotless

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -668,51 +668,57 @@ private void handleJoinRequest(JoinRequest joinRequest, ActionListener<Void> joi
-                                Releasables.close(response);
-                                joinListener.onFailure(e);
+                            // Immediate condition check in case another node is elected master
+                            if (clusterService.state().nodes().nodeExists(joinRequest.getSourceNode().getId())) {
```
> Does it work

Hmm no, not quite: that might complete `ll` twice which is generally something we should try and avoid. It turns out that as things are implemented today `ll` will be a `SubscribableListener` which has well-defined semantics when completed multiple times:

https://github.com/elastic/elasticsearch/blob/cdc74748dcfd222e5970fa27b3e284f46a2c84e7/server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java#L44-L45

However we can't rely on that being true in future, it's not guaranteed that `ll` will always be a `SubscribableListener` in this context. I'd prefer we explicitly deduplicated this work e.g. by creating another `SubscribableListener`.

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Changes

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -220,45 +197,29 @@ Therefore, this WARN log should not be thrown (#ES-11449)
-            safeAwait(publishingBanRemovedListener);
-            logger.info("Master publishing ban removed");
+            // Assert the master was re-elected
+            assertTrue(masterNodeName.equals(internalCluster().getMasterName()) && originalTerm < getTerm(masterNodeName));
```
nit: slight preference for this being two `assertThat` calls: if this fails, we get no feedback about why, whereas if we use `assertThat` twice then we'll be able to see whether it was the master that changed (and it'll identify the new master) or whether the term didn't increase.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -668,51 +668,57 @@ private void handleJoinRequest(JoinRequest joinRequest, ActionListener<Void> joi
+                                // Remove this listener to avoid memory leaks
+                                clusterService.removeListener(clusterStateListener);
+                                ll.onFailure(e);
                             }
```
It seems like it'd be simpler to call `clusterStateListener.clusterChanged` directly here, constructing a "fake" ClusterChangedEvent to carry the current cluster state. I think we want the same logic both ways: particularly if the node is for some reason in the current cluster state at this point then we can complete `ll` successfully too.

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java:
```diff
@@ -0,0 +1,257 @@
+        Node N should join the cluster, but it should not be disconnected (#ES-11449)
+     */
+    @TestLogging(reason = "test includes assertions about logging", value = "org.elasticsearch.cluster.coordination:INFO")
+    public void testNodeTriesToJoinClusterAndThenSameMasterIsElected() {
```
Btw this test is a really good discriminator of the improvement, it failed for me 9 out of 10 times I ran it having reverted the behaviour change in `Coordinator`, and passed 10 out of 10 with the change in place. Great stuff.

joshua-adams-1 committed:
David Comments

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

joshua-adams-1 committed:
Merge branch 'master-node-disconnect' of github.com:joshua-adams-1/elasticsearch into master-node-disconnect

joshua-adams-1 committed:
David Comments

joshua-adams-1 committed:
Merge branch 'main' into master-node-disconnect

DaveCTurner reviewed (approved):
LGTM
